### ISSUE TITLE
[promise-function-async] Should not fix `Promise<void> | void` functions

### ISSUE THREAD

**Issue by fregante** [typescript-eslint/typescript-eslint#3580]
There are some situations where "maybe" returning a Promise make sense, for example in Web Extensions (see the large warning section for [onMessage](https://developer.mozilla.org/en-US/docs/Mozilla/Add-ons/WebExtensions/API/runtime/onMessage#parameters))

I just run the autofix for this rule and it changed:

```ts
function handler(): Promise<string> | void {/* etc */}
```

to

```ts
async function handler(): Promise<string> | void {/* etc */}
```

Which is broken: `void` will never be the return value.

## Template

- [x] I have tried restarting my IDE and the issue persists.
- [x] I have updated to the latest version of the packages.
- [x] I have [read the FAQ](https://github.com/typescript-eslint/typescript-eslint/blob/master/docs/getting-started/linting/FAQ.md) and my problem is not listed.

**Repro**

```JSON
{
  "rules": {
    "@typescript-eslint/promise-function-async": "error"
  }
}
```

```TS
function handler(): Promise<string> | void {
	if (Math.random() > 0.9) {
		return Promise.resolve('Lucky!');
	}
}
```

**Expected Result**

- It should not try to autofix it (definitely)
- It should not mark it as a problem (maybe?)

**Actual Result**

```
1:15  error    Functions that return promises must be async          @typescript-eslint/promise-function-async
```

**Versions**

| package                            | version |
| ---------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin` | `4.27.0` |
| `@typescript-eslint/parser`        | `4.27.0` |
| `TypeScript`                       | `4.3.2` |
| `ESLint`                           | `7.28.0` |
| `node`                             | `14` |

**Comment by bradzacher** [comment]
I'm unable to repro this against master - the code you have provided does not report

![image](https://user-images.githubusercontent.com/7462525/124649915-f3db0f80-de4d-11eb-863c-78ce02784b20.png)

**Comment by fregante** [comment]
Unless it was fixed but not yet released, I can replicate after reinstalling fresh:

## Repro

[Download.zip](https://github.com/typescript-eslint/typescript-eslint/files/6772265/Archive.zip)

```
npm install -y
npm i --save-dev eslint typescript @typescript-eslint/parser @typescript-eslint/eslint-plugin
```

```js
module.exports = {
	root: true,
	parser: '@typescript-eslint/parser',
	plugins: ['@typescript-eslint'],
	rules: {
		'@typescript-eslint/promise-function-async': 'error',
	},
	parserOptions: {
		project: 'tsconfig.json',
	},
};
```

```
// tsconfig.json
{} 
```

```ts
// index.ts
export function backgroundListener(): Promise<string> {
	if (Math.random() > 0.9) {
		return Promise.resolve('Lucky!');
	}
}
```

**Comment by fregante** [comment]
```
❯ npx eslint index.ts

~/repro/index.ts
  1:8  error  Functions that return promises must be async  @typescript-eslint/promise-function-async

✖ 1 problem (1 error, 0 warnings)
  1 error and 0 warnings potentially fixable with the `--fix` option.
```

**Comment by bradzacher** [comment]
```ts
export function backgroundListener(): Promise<string> {
	if (Math.random() > 0.9) {
		return Promise.resolve('Lucky!');
	}
}
```

This code is only compile-time valid if you have both `strict` (/ `strictNullChecks`) turned off, and you also have `noImplicitReturns` turned off.

your original code samples (which annotated the return type as `Promise<string> | void`) is both compile-time valid, and ignored by the lint rule.

---

For your new sample, we can't actually tell that your code returns `undefined` in addition to a `Promise` - the `undefined` case is entirely removed from the types.

**Comment by fregante** [comment]
My last example didn't include the core of the issue: `| undefined`. Please review again because I don't follow your reasoning. The code is perfectly valid and the rule is incorrectly asking my to change the type of the function while it can't.

Here's a better example:

```ts
export function backgroundListener(): Promise<string> | string {
	if (Math.random() > 0.9) {
		return Promise.resolve('Lucky!');
	}

	return ''
}
```

But focus on the function header and ignore the body:

```ts
export function backgroundListener(): Promise<string> | string
```

This function cannot be turned into an async function.

**Comment by bradzacher** [comment]
So to clarify. The following cases are ignored by the lint rule:

- `function foo(): Promise<string> | string`
- `function foo(): Promise<string> | void`
- `function foo(): Promise<string> | undefined`
- `function foo(): Promise<string> | null`

![image](https://user-images.githubusercontent.com/7462525/124666426-e1b79c00-de62-11eb-8ca1-b2dd5f27d3e6.png)

The only case you've presented that is not ignored by the lint rule is:
`function foo(): Promise<string>`

From the zip file you uploaded, the function that is erroring is this:
```ts
export function backgroundListener(): Promise<string> {
	if (Math.random() > 0.9) {
		return Promise.resolve('Lucky!');
	}
}
```

So.. why does this code error?

Well in its current form - the lint rule simply looks at the function signature.
As far as we can tell from the function's signature - it's just returning a `Promise<string>`. There is no `void`/`undefined` in the return type signature.
So as far as our static analysis can tell - it's just a plain old async function and should be marked as such.

But wait! When we inspect the function's body - we can very clearly see that this function has two branches - one that returns `Promise<string>`, and one that returns nothing (`void`).

So why is this code TS valid?!?
Well in your zip the `tsconfig` you provided was just `{}` - which means "use the default for all the options". By default all "strict" compiler options are off, so `strictNullChecks` is off.
With `strictNullChecks` off - TS allows this loose and bad code where you're returning nothing from a function that declares it always returns a `Promise`.

So what can this lint rule do about this?
Well there's only one way to handle it - do complete control-flow analysis of the function body to determine if it ever returns any of `null | undefined | void`...
This is a **lot** of work that is going to be *very* slow, and honestly it's not worth it when you (a) should be using `strictNullChecks`, and (b) can fix this by correctly annotating your function

**Comment by fregante** [comment]
Once again, my example did not include the `| undefined` part and since that part is apparently problematic I offered another example. Please refer to that and ignore `undefined`/`void`. This has nothing to do with null checks.

**Comment by fregante** [comment]
I'll open a new issue to be more clear.

**Comment by bradzacher** [comment]
I'm confused - I thought I addressed all cases with my response.

Unions with `export function backgroundListener(): Promise<string> | string` is erroring for you, but I cannot reproduce that (see screenshots of test output against master).

The union case is explicitly ignored in the lint rule due to this code here:

https://github.com/typescript-eslint/typescript-eslint/blob/e3a3ea04757464aa2dded1ef46af8ad4e05246f2/packages/eslint-plugin/src/rules/promise-function-async.ts#L89-L92

https://github.com/typescript-eslint/typescript-eslint/blob/e3a3ea04757464aa2dded1ef46af8ad4e05246f2/packages/eslint-plugin/src/rules/promise-function-async.ts#L112-L121

https://github.com/typescript-eslint/typescript-eslint/blob/cae4f4a0f33f8c954b1670d0abcfc8edd6193a06/packages/eslint-plugin/src/util/types.ts#L64-L68

Every type in the union must be a promise type for the rule to fire

**Comment by fregante** [comment]
I'll just say this:

- The repro.zip, once I add `| undefined` and `return undefined`, reports an error
- If I use `| void`, it no longer reports an error
- If I use a [strict tsconfig](https://github.com/sindresorhus/tsconfig/blob/main/tsconfig.json), it no longer reports an error

I'm not sure why my IDE was reporting an error even with `Promise<string> | string` but I can't make it happen anymore 😅 

Either way I have a solution for my case (generally it's a stricter config, but in this case it should have been `void`)

I still think it shouldn't report the first case (loose config, `| undefined`), but if you disagree that's ok too 😁

**Comment by bradzacher** [comment]
Ah yes. `| void`, `| undefined` and `| null` with `strictNullChecks` off.

TL;DR is that we cannot handle that case.
When `strictNullChecks` is off TS essentially "erases" the nullish types.
So all that our tooling sees is `Promise<string>` - we don't see that you declared a nullish type union!

This is the same reason that both [`no-unnecessary-condition`](https://github.com/typescript-eslint/typescript-eslint/blob/master/packages/eslint-plugin/docs/rules/no-unnecessary-condition.md#allowruletorunwithoutstrictnullchecksiknowwhatiamdoing) and [`strict-boolean-expressions`](https://github.com/typescript-eslint/typescript-eslint/blob/master/packages/eslint-plugin/docs/rules/strict-boolean-expressions.md#allowruletorunwithoutstrictnullchecksiknowwhatiamdoing) don't actually work without `strictNullChecks` turned on.
It's just how TS's type system works unfortunately.

**Comment by fregante** [comment]
> So all that our tooling sees is `Promise<string>` - we don't see that you declared a nullish type union!

Ohhh that's the piece of information that wasn't coming through. I didn't know about that limitation
